src/main/java/br/com/caelum/viagens/voos/validator/RotaExistenteValidator.java

```diff
@@ -0,0 +1,40 @@
+package br.com.caelum.viagens.voos.validator;
+
+import java.util.List;
+import java.util.Optional;
+import java.util.stream.Collectors;
+
+import org.springframework.validation.Errors;
+import org.springframework.validation.Validator;
+
+import br.com.caelum.viagens.administrativo.repository.RotaRepository;
+import br.com.caelum.viagens.voos.controller.dto.input.NewVooInputDto;
+
+public class RotaExistenteValidator implements Validator {
+
+	private RotaRepository rotaRepository;
+
+	public RotaExistenteValidator(RotaRepository rotaRepository) {
+		this.rotaRepository = rotaRepository;
+	}
+
+	@Override
+	public boolean supports(Class<?> clazz) {
+		return NewVooInputDto.class.isAssignableFrom(clazz);
+	}
+
+	@Override
+	public void validate(Object target, Errors errors) {
+		NewVooInputDto newVooDto = (NewVooInputDto) target;
+
+		List<Optional<br.com.caelum.viagens.administrativo.model.Rota>> rotas = newVooDto.getRotas().stream()
```

Comment:
Um outro jeito de você executar a validaçao aqui, com menos queries, era criando um método que recebe uma coleção de ids de rotas e retorna uma coleção de rotas. Se o número de rotas retornadas for diferente do número passado, está inváldo. Só sugestão

Reply:
eu modifiquei o código para executar apenas 1 query como você sugeriu. mas não validei comparando os tamanhos das coleções já que entraria em conflito com o validador de rotas repetidas.